### ISSUE TITLE
Unreliable console keepalive

### ISSUE THREAD

**Issue by dfortner** [hercules-390/hyperion#44]
David H. Durgee reports:

<blockquote>
I have recently been doing a bit more testing with hyperion built from a pull on 2014/03/25 on my linux mint 13 maya x64 system.  I noticed that despite conkpalv set as 3,1,10 my 3270 console sessions would hang without activity over time.  Wireshark shows that no keepalive packets are being exchanged for the 3270 connections.  Interestingly, wireshark shows a telnet session to a 2703 line DOES have keepalive packets exchanged.  Strange.
</blockquote>

The fact that it sometimes works (for 2703 devices) but at other times does not (for 3270 console devices) would seem to provide strong evidence that Hyperion's TCP keepalive is in some way broken (at least for 3270 console devices anyway).

In any case this issue definitely merits further research.

**Comment by BugAndNewsReporter** [comment]
I'm going to add some additional debugging/reporting code to Hercules that will hopefully report what the heck is going on, so just hang loose for now. I'll post the patch here when it's ready. Thanks.

**Comment by Fish-Git** [comment]
(ATTACHMENT)
(off list)
(CC: github issue #44)

Fish wrote:
> dhdurgee@verizon.net wrote:
> 
> > I don't see such a file on my linux system. [...]
> 
> And that may indeed be at the root of the problem, but
> let's be sure about it. Let me add some debugging code
> to Hercules first so we can see what's going on with
> your system (and to detect the issue in case it happens
> on others' systems too).
> 
> I'll let you know when it's ready.

Dave,

It's ready. Sorry it took me so long.

Please apply the attached patch and rebuild Hercules (sh ./autogen.sh && ./configure && make) and then let me know whether it resolves your issue or not.

I've added some new warning messages to not only the build process itself (to alert you if the headers required for keepalive support cannot be found on your system (and thus keepalive support cannot even be built)), but also to Hercules's startup and initialization logic as well as to its 'conkpalv' command too, to alert you to any problems encountered during its effort to enable keepalive at runtime.

Be sure to rebuild your configure script (sh ./autogen.sh) and re-configure Hercules (./configure) before rebuilding Hercules. I had to make some changes to the build process to correct what I believe was at the root of the problem(*).

Try both with/without a "conkpalv" statement in your configuration file as well as using the "conkpalv" panel command to dynamically change/display the current values.


PLEASE post a summary of your test results to the GitHub issue I created for you:

  https://github.com/hercules-390/hyperion/issues/44


You can reply to me directly if you want but I'd prefer that you do it in the GitHub issue.

Thanks.

-- 
"Fish" (David B. Trout)
Software Development Laboratories
http://www.softdevlabs.com
mail: fish@softdevlabs.com


(*) Not looking for the tcp.h header in the right place as well as other configure.ac issues.

**Comment by dhdurgee** [comment]
I applied your patch and rebuilt my hyperion pull as directed.  Unfortunately I am seeing no difference in behavior from before applying the patch.  Perhaps I am missing something on my system?  I kept logs of the patch and various steps of the reconfiguration and build if that would be helpful to you.  Just let me know what would be useful in diagnosis.  Alternatively, if there are some additional options that I could define and make another stab at the build with please let me know.

**Comment by Fish-Git** [comment]
Please send me all four logs: autogen, configure, make and hercules.

Thanks.

-- 
"Fish" (David B. Trout)
Software Development Laboratories
http://www.softdevlabs.com
mail: fish@softdevlabs.com

**Comment by Fish-Git** [comment]
Also, just out of curiosity, what does the following command report on your system?

    find  /  -iname  'tcp.h'

Thanks.

-- 
"Fish" (David B. Trout)
Software Development Laboratories
http://www.softdevlabs.com
mail: fish@softdevlabs.com

**Comment by Fish-Git** [comment]
Fish wrote:

> Also, just out of curiosity, what does the following command
> report on your system?
> 
>     find  /  -iname  'tcp.h'
> 
> Thanks.

Er, make that:

    find / -iname 'tcp.h' 2>&1 | grep "tcp.h"

instead. Thanks.

**Comment by Fish-Git** [comment]
Fish wrote:

> Please send me all four logs: autogen, configure,
> make and hercules.

Also send me your "config.h" header file too.

Thanks.

**Comment by jphartmann** [comment]
Perhaps this might be better:

find /usr -name tcp.h -ls

**Comment by dhdurgee** [comment]
Fish-Git wrote:
> Fish wrote:
>
> > Please send me all four logs: autogen, configure,
> > make and hercules.
>
> Also send me your "config.h" header file too.
>
> Thanks.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/hercules-390/hyperion/issues/44#issuecomment-70057313>.
>
As I can't attach a .zip file on github I am responding here. Attached 
please find what you asked for,  If you need anything else, or if I 
somehow overlooked something, please let me know and I will send 
additional data.

Dave

**Comment by Fish-Git** [comment]
dhdurgee wrote:
> Fish-Git wrote:
> > Fish wrote:
> >
> > > Please send me all four logs: autogen, configure,
> > > make and hercules.
> >
> > Also send me your "config.h" header file too.
> >
> > Thanks.
> >
> As I can't attach a .zip file on github I am responding here.

Where is "here"?

Because I am responding via github issue notification email reply as it was the only response from you that I received. I did not receive any other message from you via any other avenue.


> Attached please find what you asked for, If you need anything
> else, or if I somehow overlooked something, please let me know
> and I will send additional data.

I did not receive any attachments since, as you stated (and as we already know), GitHub does not support attachments (which is where I am responding from since this is the only place where I received any email from).

Even though I did not specifically say so in my original request, I was hoping you would send the files to me directly. My email address is in my signature.

Thanks.

-- 
"Fish" (David B. Trout)
Software Development Laboratories
http://www.softdevlabs.com
mail: fish@softdevlabs.com

**Comment by Fish-Git** [comment]
(ATTACHMENT)
CC: github

David H. Durgee wrote:
[...]
> PS: Opps, I thought the github limit only applied to
> the GUI input.  I am resending to your email.  I will
> try to remember to CC it next time.

Thanks. Got them.

I took a peek but could see nothing blatantly unusual(*).

It's late and I'm tired though. Will start fresh tomorrow.

Attached is a little test program we may need to modify to get to the bottom of what's going on. Do me a favor and give it a try.

Make sure the "server" runs on a different system than the "client", and then let the client sit (go idle) and then check if keepalives are occurring.(**)

Thanks. I'll be in touch.


--------------
(*) Besides the weird "unused result" gcc warning which I've never seen before! What version of GCC are you using?

(**) How do you do that on Linux? On Windows I use Wireshark. Does Linux have Wireshark?

-- 
"Fish" (David B. Trout)
Software Development Laboratories
http://www.softdevlabs.com
mail: fish@softdevlabs.com


//  http://www.tldp.org/HOWTO/html_single/TCP-Keepalive-HOWTO/
//  http://www.binarytides.com/server-client-example-c-sockets-linux/


#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <arpa/inet.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <string.h>

#define STR(s)          #s
#define XSTR(s)         STR(s)
#define MAX_MSG_LEN     2000

static int syntax()
{
    printf(

        "\nSimple client/server TCP keepalive test\n\n"

        "Format:\n\n"

        "   keeptest   { \"server\" port} | { \"client\" ip port }\n\n"

        "Where:\n\n"

        "    ip        is the ip address where the server is running\n"
        "    port      is the port number the server is listening on\n\n"

        "The \"server\" option starts a listening server on the specified\n"
        "port. The \"client\" option connects to the server at the specified\n"
        "ip address and port number.\n\n"

        "The test is a simple echo test where everything entered on the client\n"
        "is sent to the server which simply echos back to the client everything\n"
        "sent to the server.\n\n"

        "The test ends when 'Ctrl+C' is entered on the client which also ends\n"
        "the server.\n\n"
    );

    return EXIT_FAILURE;
}

/* Enable and report basic keepalive support */

static int keepalive( int socket )
{
    int optval;
    socklen_t optlen = sizeof( optval );

    /* Check the status for the keepalive option */

    if (getsockopt( socket, SOL_SOCKET, SO_KEEPALIVE, &optval, &optlen ) < 0)
    {
        perror("getsockopt()");
        return EXIT_FAILURE;
    }

    printf("SO_KEEPALIVE is %s\n", (optval ? "ON" : "OFF"));

    /* Set the option active */

    optval = 1;
    optlen = sizeof(optval);

    if (setsockopt( socket, SOL_SOCKET, SO_KEEPALIVE, &optval, optlen ) < 0)
    {
        perror("setsockopt()");
        return EXIT_FAILURE;
    }

    printf("SO_KEEPALIVE set on socket\n");

    /* Check the status again */

    if (getsockopt( socket, SOL_SOCKET, SO_KEEPALIVE, &optval, &optlen ) < 0)
    {
        perror("getsockopt()");
        return EXIT_FAILURE;
    }

    printf("SO_KEEPALIVE is now %s\n", (optval ? "ON" : "OFF"));

    return optval ? EXIT_SUCCESS : EXIT_FAILURE;
}

/* Echoing socket server */

int server( int listen_port )
{
    int  listen_socket;
    int  client_socket;
    int  read_size;
    int  size;

    struct sockaddr_in  server;
    struct sockaddr_in  client;

    char  client_message[ MAX_MSG_LEN ];

    // Create listening socket

    listen_socket = socket( AF_INET, SOCK_STREAM, 0 );

    if (listen_socket == -1)
    {
        perror("socket()");
        return EXIT_FAILURE;
    }
    printf("Socket created\n");

    // Prepare the sockaddr_in structure

    server.sin_family       =  AF_INET;
    server.sin_addr.s_addr  =  INADDR_ANY;
    server.sin_port         =  htons( listen_port );

    // Bind

    if (bind( listen_socket, (struct sockaddr*) &server, sizeof( server )) < EXIT_SUCCESS)
    {
        perror("bind()");
        close( listen_socket );
        return EXIT_FAILURE;
    }
    printf("bind done\n");

    // Listen for connection

    listen( listen_socket, 3 );

    // Accept any incoming connection

    printf("Waiting for incoming connection to port %d...\n", listen_port);

    size = sizeof( struct sockaddr_in );

    // accept connection from an incoming client

    client_socket = accept( listen_socket, (struct sockaddr*) &client, (socklen_t*) &size );

    if (client_socket < EXIT_SUCCESS)
    {
        perror("accept()");
        close( listen_socket );
        return EXIT_FAILURE;
    }

    printf("Connection accepted!\n");

    close( listen_socket ); // (no longer needed)

    // Enable keepalive on client socket

    if (keepalive( client_socket ) != EXIT_SUCCESS)
    {
        printf("Enable keepalive failed\n");
        close( client_socket );
        return EXIT_FAILURE;
    }

    // Loop until client disconnects or an error occurs

    while (1)
    {
        // Read message from client

        if ((read_size = recv( client_socket, client_message, MAX_MSG_LEN, 0 )) <= 0)
        {
            perror("recv()");
            break;
        }
        
        if (read_size >= MAX_MSG_LEN)
            read_size = MAX_MSG_LEN - 1;

        client_message[ read_size ] = 0; // (null terminate)
        size = strlen( client_message );

        // Send the message back to client

        if (write( client_socket, client_message, size ) != size)
        {
            perror("write()");
            break;
        }
    }

    close( client_socket );

    if (read_size == 0)
    {
        printf("Client disconnected. Test ended.\n");
        fflush( stdout );
    }

    return EXIT_SUCCESS;
}

/* Echoing socket client */

int client( char* server_ip, int server_port )
{
    int  rc, sock, msglen;

    struct sockaddr_in  server;

    char client_message[ MAX_MSG_LEN ];
    char server_message[ MAX_MSG_LEN ];

    // Create socket

    sock = socket( AF_INET, SOCK_STREAM, 0 );
    if (sock == -1)
    {
        perror("socket()");
        return EXIT_FAILURE;
    }

    printf("Socket created\n");

    server.sin_family       =  AF_INET;
    server.sin_addr.s_addr  =  inet_addr( server_ip );
    server.sin_port         =  htons( server_port );

    // Connect to remote server

    if (connect( sock, (struct sockaddr*) &server, sizeof( server )) < EXIT_SUCCESS)
    {
        perror("connect()");
        return EXIT_FAILURE;
    }

    printf("Connected! Use Ctrl+C to end test.\n");

    // keep communicating with server

    rc = EXIT_SUCCESS;

    while (1)
    {
        printf("\nEnter message (%d chars max):\n\n", MAX_MSG_LEN - 1);

        client_message[0] = 0; // (empty buffer)
        scanf( "%" XSTR( MAX_MSG_LEN ) "[^\n]s", client_message );
        getchar(); // (read newline)
        client_message[  MAX_MSG_LEN - 1  ] = 0; // (null terminate)
        msglen = strlen( client_message );
        
        // Send the message to the server

        if (send( sock, client_message, msglen, 0 ) != msglen)
        {
            perror("send()");
            rc = EXIT_FAILURE;
            break;
        }

        // Read the reply from the server

        server_message[0] = 0; // (empty buffer)
        
        if ((msglen = recv( sock, server_message, MAX_MSG_LEN, 0 )) < 0)
        {
            perror("recv()");
            rc = EXIT_FAILURE;
            break;
        }

        // Check for server disconnect

        if (!msglen)
            break;

        if (msglen >= MAX_MSG_LEN)
            msglen = MAX_MSG_LEN - 1;

        server_message[ msglen ] = 0;
        printf("\nServer reply:\n\n");
        printf("%s\n", server_message);
    }

    close( sock );

    return rc;
}

int main( int argc, char *argv[] )
{
    int server_port;
    char* server_ip;

    if (argc < 2)
        return syntax();

    if (strcasecmp( argv[1], "server" ) == 0)
    {
        if (argc != 3)
            return syntax();

        server_port = atoi( argv[2] );

        if (!server_port)
            return syntax();

        return server( server_port );
    }
    else if (strcasecmp( argv[1], "client" ) == 0)
    {
        if (argc != 4)
            return syntax();

        server_ip = argv[2];
        server_port = atoi( argv[3] );

        if (!server_port)
            return syntax();

        return client( server_ip, server_port );
    }
    else
        return syntax();
}
=

**Comment by Fish-Git** [comment]
(attachments)
(CC: github)

David H. Durgee wrote:

> I decided to take a simple stab at this and find
> that gcc keeptest.c created a working a.out on my
> server system which runs on both systems.
> 
> Given this I gave it a try for you.  Attached please
> find a zip file containing logs from both server and
> client side and a wireshark capture file of traffic
> between the two systems.  As you can see no keepalive
> packets were exchanged for this connection.

Bad test. My fault. My keeptest.c was too simplistic and only enabled "basic" (default) keepalive, which defaults to 2 hours. (Hey I told you I was tired!)

Try the attached new/improved version instead. It contains the same logic as in that original Hercules patch I sent you (conkpalv.patch), plus a tiny bit more to report the actual keepalive values being used.

Thanks.

-- 
"Fish" (David B. Trout)
Software Development Laboratories
http://www.softdevlabs.com
mail: fish@softdevlabs.com


//  http://www.tldp.org/HOWTO/html_single/TCP-Keepalive-HOWTO/
//  http://www.binarytides.com/server-client-example-c-sockets-linux/


#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <arpa/inet.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <string.h>
#include <errno.h>
#include <netdb.h>
#include <netinet/tcp.h>

// (simulate config.h and hercules.h headers)

#include "keeptest.h"

// Additional handy macros that we use...

#define STR(s)          #s
#define XSTR(s)         STR(s)
#define MAX_MSG_LEN     2000

// Simulate Hercules SYSBLK

SYSBLK sysblk = {0};

static int syntax()
{
    printf(

        "\nSimple client/server TCP keepalive test\n\n"

        "Format:\n\n"

        "   keeptest   { \"server\" port} | { \"client\" ip port }\n\n"

        "Where:\n\n"

        "    ip        is the ip address where the server is running\n"
        "    port      is the port number the server is listening on\n\n"

        "The \"server\" option starts a listening server on the specified\n"
        "port. The \"client\" option connects to the server at the specified\n"
        "ip address and port number.\n\n"

        "The test is a simple echo test where everything entered on the client\n"
        "is sent to the server which simply echos back to the client everything\n"
        "sent to the server.\n\n"

        "The test ends when 'Ctrl+C' is entered on the client which also ends\n"
        "the server.\n\n"
    );

    return EXIT_FAILURE;
}

/* Function to retrieve keepalive values. 0==success, -1=failure */
int get_socket_keepalive( int sfd, int* idle_time, int* probe_interval,
                          int* probe_count )
{
    struct protoent*  tcpproto;
    int        optval, l_tcp;
    socklen_t  optlen = sizeof( optval );

    /* Retrieve TCP protocol value (mostly for FreeBSD portability) */
    tcpproto = getprotobyname("TCP");
    if (!tcpproto)
    {
        WRMSG( HHC02219, "E", "getprotobyname(\"TCP\")", strerror( HSO_errno ));
        return -1;
    }
    l_tcp = tcpproto->p_proto;

    /* Default values */
    *idle_time      = sysblk.kaidle;
    *probe_interval = sysblk.kaintv;
    *probe_count    = sysblk.kacnt;

#if defined( HAVE_FULL_KEEPALIVE )
    /* Retrieve the actual values from the system */
    if (getsockopt( sfd, l_tcp, TCP_KEEPIDLE,  &optval, &optlen ) >= 0) *idle_time      = optval;
    if (getsockopt( sfd, l_tcp, TCP_KEEPINTVL, &optval, &optlen ) >= 0) *probe_interval = optval;
    if (getsockopt( sfd, l_tcp, TCP_KEEPCNT,   &optval, &optlen ) >= 0) *probe_count    = optval;
#endif // HAVE_FULL_KEEPALIVE

    return 0;
}

/* Set the SO_KEEPALIVE option and timeout values for a
   socket connection to detect when client disconnects.
   Returns 0==success, +1==warning, -1==failure
   (*) Warning failure means function only partially
       succeeded (not all requested values were set)
*/
int set_socket_keepalive( int sfd,
                          int idle_time,
                          int probe_interval,
                          int probe_count )
{
#if !defined( HAVE_FULL_KEEPALIVE ) && !defined( HAVE_BASIC_KEEPALIVE )
    UNREFERENCED( sfd );
    UNREFERENCED( idle_time );
    UNREFERENCED( probe_interval );
    UNREFERENCED( probe_count );
    return -1;
#else // defined( HAVE_BASIC_KEEPALIVE ) || defined( HAVE_FULL_KEEPALIVE )

    int idle, intv, cnt;
    int rc, l_tcp, optval, succeeded = 0, tried = 0;
    struct protoent * tcpproto;

    /* Retrieve TCP protocol value (mostly for FreeBSD portability) */
    tcpproto = getprotobyname("TCP");
    if (!tcpproto)
    {
        WRMSG( HHC02219, "E", "getprotobyname(\"TCP\")", strerror( HSO_errno ));
        return -1;
    }
    l_tcp = tcpproto->p_proto;

    optval = 1;
  #if defined( HAVE_DECL_SO_KEEPALIVE ) && HAVE_DECL_SO_KEEPALIVE
    tried++;
    rc = setsockopt(sfd, l_tcp, SO_KEEPALIVE, &optval, sizeof(optval));
    if (rc)
    {
        WRMSG( HHC02219, "E", "setsockopt(SO_KEEPALIVE)", strerror( HSO_errno ));
        return -1;
    }
    else
        succeeded++;
  #endif

    optval = idle_time;
  #if defined( HAVE_DECL_TCP_KEEPALIVE ) && HAVE_DECL_TCP_KEEPALIVE
    tried++;
    rc = setsockopt(sfd, IPPROTO_TCP, TCP_KEEPALIVE, &optval, sizeof(optval));
    if (rc)
        WRMSG( HHC02219, "E", "setsockopt(TCP_KEEPALIVE)", strerror( HSO_errno ));
    else
        succeeded++;
  #endif

    optval = idle_time;
  #if defined( HAVE_DECL_TCP_KEEPIDLE ) && HAVE_DECL_TCP_KEEPIDLE
    tried++;
    rc = setsockopt(sfd, l_tcp, TCP_KEEPIDLE, &optval, sizeof(optval));
    if (rc)
        WRMSG( HHC02219, "E", "setsockopt(TCP_KEEPIDLE)", strerror( HSO_errno ));
    else
        succeeded++;
  #endif

    optval = probe_interval;
  #if defined( HAVE_DECL_TCP_KEEPINTVL ) && HAVE_DECL_TCP_KEEPINTVL
    tried++;
    rc = setsockopt(sfd, l_tcp, TCP_KEEPINTVL, &optval, sizeof(optval));
    if (rc)
        WRMSG( HHC02219, "E", "setsockopt(TCP_KEEPINTVL)", strerror( HSO_errno ));
    else
        succeeded++;
  #endif

    optval = probe_count;
  #if defined( HAVE_DECL_TCP_KEEPCNT ) && HAVE_DECL_TCP_KEEPCNT
    tried++;
    rc = setsockopt(sfd, l_tcp, TCP_KEEPCNT, &optval, sizeof(optval));
    if (rc)
        WRMSG( HHC02219, "E", "setsockopt(TCP_KEEPCNT)", strerror( HSO_errno ));
    else
        succeeded++;
  #endif

    idle = idle_time;
    intv = probe_interval;
    cnt  = probe_count;

    /* Retrieve values in use, ignoring any error */
    get_socket_keepalive( sfd, &idle, &intv, & cnt );

    /* Determine return code: did everything succeed and,
       more importantly, were all of our values accepted?
       (i.e. did they all "stick"?)
    */
    if (succeeded >= tried)
        rc = (idle != idle_time || intv != probe_interval || cnt != probe_count) ? +1 : 0;
    else
        rc = (succeeded <= 0 ? -1 : +1);

    return rc;

#endif // KEEPALIVE
}

/* Echoing socket server */

int server( int listen_port )
{
    int  listen_socket;
    int  client_socket;
    int  read_size;
    int  size;

    struct sockaddr_in  server;
    struct sockaddr_in  client;

    char  client_message[ MAX_MSG_LEN ];

    // Create listening socket

    listen_socket = socket( AF_INET, SOCK_STREAM, 0 );

    if (listen_socket == -1)
    {
        perror("socket()");
        return EXIT_FAILURE;
    }
    printf("Socket created\n");

    // Prepare the sockaddr_in structure

    server.sin_family       =  AF_INET;
    server.sin_addr.s_addr  =  INADDR_ANY;
    server.sin_port         =  htons( listen_port );

    // Bind

    if (bind( listen_socket, (struct sockaddr*) &server, sizeof( server )) < EXIT_SUCCESS)
    {
        perror("bind()");
        close( listen_socket );
        return EXIT_FAILURE;
    }
    printf("bind done\n");

    // Listen for connection

    listen( listen_socket, 3 );

    // Accept any incoming connection

    printf("Waiting for incoming connection to port %d...\n", listen_port);

    size = sizeof( struct sockaddr_in );

    // accept connection from an incoming client

    client_socket = accept( listen_socket, (struct sockaddr*) &client, (socklen_t*) &size );

    if (client_socket < EXIT_SUCCESS)
    {
        perror("accept()");
        close( listen_socket );
        return EXIT_FAILURE;
    }

    printf("Connection accepted!\n");

    close( listen_socket ); // (no longer needed)

    //-----------------------------------------------------------------------------------
    printf("Setting keepalive on client socket...\n");
    {
        int rc, idle, intv, cnt;
#if 0
        set_socket_keepalive( client_socket, sysblk.kaidle, sysblk.kaintv, sysblk.kacnt );
#else
        rc = set_socket_keepalive( client_socket, sysblk.kaidle, sysblk.kaintv, sysblk.kacnt );
        if (rc < 0)
            WRMSG( HHC02219, "E", "set_socket_keepalive()", strerror( HSO_errno ));
        else if (rc > 0)
            WRMSG( HHC02320, "W" );
#endif
        rc = get_socket_keepalive( client_socket, &idle, &intv, &cnt );
        if (rc < 0)
            WRMSG( HHC02219, "E", "get_socket_keepalive()", strerror( HSO_errno ));
        else
        {
            printf("KEEPALIVE: (%d,%d,%d)\n", idle, intv, cnt );
        }
    }

    //-----------------------------------------------------------------------------------

    // Loop until client disconnects or an error occurs

    while (1)
    {
        // Read message from client

        if ((read_size = recv( client_socket, client_message, MAX_MSG_LEN, 0 )) <= 0)
        {
            perror("recv()");
            break;
        }
        
        if (read_size >= MAX_MSG_LEN)
            read_size = MAX_MSG_LEN - 1;

        client_message[ read_size ] = 0; // (null terminate)
        size = strlen( client_message );

        // Send the message back to client

        if (write( client_socket, client_message, size ) != size)
        {
            perror("write()");
            break;
        }
    }

    close( client_socket );

    if (read_size == 0)
    {
        printf("Client disconnected. Test ended.\n");
        fflush( stdout );
    }

    return EXIT_SUCCESS;
}

/* Echoing socket client */

int client( char* server_ip, int server_port )
{
    int  rc, sock, msglen;

    struct sockaddr_in  server;

    char client_message[ MAX_MSG_LEN ];
    char server_message[ MAX_MSG_LEN ];

    // Create socket

    sock = socket( AF_INET, SOCK_STREAM, 0 );
    if (sock == -1)
    {
        perror("socket()");
        return EXIT_FAILURE;
    }

    printf("Socket created\n");

    server.sin_family       =  AF_INET;
    server.sin_addr.s_addr  =  inet_addr( server_ip );
    server.sin_port         =  htons( server_port );

    // Connect to remote server

    if (connect( sock, (struct sockaddr*) &server, sizeof( server )) < EXIT_SUCCESS)
    {
        perror("connect()");
        return EXIT_FAILURE;
    }

    printf("Connected! Use Ctrl+C to end test.\n");

    // keep communicating with server

    rc = EXIT_SUCCESS;

    while (1)
    {
        printf("\nEnter message (%d chars max):\n\n", MAX_MSG_LEN - 1);

        client_message[0] = 0; // (empty buffer)
        scanf( "%" XSTR( MAX_MSG_LEN ) "[^\n]s", client_message );
        getchar(); // (read newline)
        client_message[  MAX_MSG_LEN - 1  ] = 0; // (null terminate)
        msglen = strlen( client_message );
        
        // Send the message to the server

        if (send( sock, client_message, msglen, 0 ) != msglen)
        {
            perror("send()");
            rc = EXIT_FAILURE;
            break;
        }

        // Read the reply from the server

        server_message[0] = 0; // (empty buffer)
        
        if ((msglen = recv( sock, server_message, MAX_MSG_LEN, 0 )) < 0)
        {
            perror("recv()");
            rc = EXIT_FAILURE;
            break;
        }

        // Check for server disconnect

        if (!msglen)
            break;

        if (msglen >= MAX_MSG_LEN)
            msglen = MAX_MSG_LEN - 1;

        server_message[ msglen ] = 0;
        printf("\nServer reply:\n\n");
        printf("%s\n", server_message);
    }

    close( sock );

    return rc;
}

int main( int argc, char *argv[] )
{
    int server_port;
    char* server_ip;





    /* Set default TCP keepalive values */

#if !defined( HAVE_BASIC_KEEPALIVE ) && !defined( HAVE_FULL_KEEPALIVE )

    WARNING("TCP keepalive headers not found; check configure.ac")
    WARNING("TCP keepalive support will NOT be generated")

    // "This build of Hercules does not support TCP keepalive"
    WRMSG( HHC02321, "E" );

#else // defined( HAVE_BASIC_KEEPALIVE ) || defined( HAVE_FULL_KEEPALIVE )

  #if defined( HAVE_BASIC_KEEPALIVE ) && !defined( HAVE_FULL_KEEPALIVE )

    WARNING("This build of Hercules will only have basic TCP keepalive support")

    // "This build of Hercules has only basic TCP keepalive support"
    WRMSG( HHC02322, "W" );

  #endif // defined( HAVE_BASIC_KEEPALIVE )
    /*
    **  Note: we need to try setting them to our desired values first
    **  and then retrieve the set values afterwards to detect systems
    **  which do not allow some values to be changed to ensure SYSBLK
    **  gets initialized with proper working default values.
    */
    {
        int rc, sfd, idle, intv, cnt;

        /* Need temporary socket for setting/getting */
        sfd = socket( AF_INET, SOCK_STREAM, 0 );
        if (sfd < 0)
        {
            WRMSG( HHC02219, "E", "socket()", strerror( HSO_errno ));
            idle = 0;
            intv = 0;
            cnt  = 0;
        }
        else
        {
            idle = KEEPALIVE_IDLE_TIME;
            intv = KEEPALIVE_PROBE_INTERVAL;
            cnt  = KEEPALIVE_PROBE_COUNT;

            /* First, try setting the desired values */
            rc = set_socket_keepalive( sfd, idle, intv, cnt );

            if (rc < 0)
            {
                WRMSG( HHC02219, "E", "set_socket_keepalive()", strerror( HSO_errno ));
                idle = 0;
                intv = 0;
                cnt  = 0;
            }
            else
            {
                /* Report partial success */
                if (rc > 0)
                {
                    // "Not all TCP keepalive settings honored"
                    WRMSG( HHC02320, "W" );

                    /* Retrieve current values from system */
                    if (get_socket_keepalive( sfd, &idle, &intv, &cnt ) < 0)
                        WRMSG( HHC02219, "E", "get_socket_keepalive()", strerror( HSO_errno ));
                }
            }
            close_socket( sfd );
        }

        /* Initialize SYSBLK with default values */
        sysblk.kaidle = idle;
        sysblk.kaintv = intv;
        sysblk.kacnt  = cnt;
    }

#endif // KEEPALIVE





    if (argc < 2)
        return syntax();

    if (strcasecmp( argv[1], "server" ) == 0)
    {
        if (argc != 3)
            return syntax();

        server_port = atoi( argv[2] );

        if (!server_port)
            return syntax();

        return server( server_port );
    }
    else if (strcasecmp( argv[1], "client" ) == 0)
    {
        if (argc != 4)
            return syntax();

        server_ip = argv[2];
        server_port = atoi( argv[3] );

        if (!server_port)
            return syntax();

        return client( server_ip, server_port );
    }
    else
        return syntax();
}

=

#ifndef _KEEPTEST_H_
#define _KEEPTEST_H_

// Simulate config.h header...

#ifdef            SO_KEEPALIVE
#define HAVE_DECL_SO_KEEPALIVE      1
#else
#define HAVE_DECL_SO_KEEPALIVE      0
#endif

#ifdef            TCP_KEEPALIVE
#define HAVE_DECL_TCP_KEEPALIVE     1
#else
#define HAVE_DECL_TCP_KEEPALIVE     0
#endif

#ifdef            TCP_KEEPIDLE
#define HAVE_DECL_TCP_KEEPIDLE      1
#else
#define HAVE_DECL_TCP_KEEPIDLE      0
#endif

#ifdef            TCP_KEEPINTVL
#define HAVE_DECL_TCP_KEEPINTVL     1
#else
#define HAVE_DECL_TCP_KEEPINTVL     0
#endif

#ifdef            TCP_KEEPCNT
#define HAVE_DECL_TCP_KEEPCNT       1
#else
#define HAVE_DECL_TCP_KEEPCNT       0
#endif

#if (defined( HAVE_DECL_SO_KEEPALIVE  ) && HAVE_DECL_SO_KEEPALIVE ) || \
    (defined( HAVE_DECL_TCP_KEEPALIVE ) && HAVE_DECL_TCP_KEEPALIVE)
#define HAVE_BASIC_KEEPALIVE
#endif

#if (defined( HAVE_DECL_TCP_KEEPIDLE ) && HAVE_DECL_TCP_KEEPIDLE ) && \
    (defined( HAVE_DECL_TCP_KEEPINTVL) && HAVE_DECL_TCP_KEEPINTVL) && \
    (defined( HAVE_DECL_TCP_KEEPCNT  ) && HAVE_DECL_TCP_KEEPCNT  )
#define HAVE_FULL_KEEPALIVE
#endif

/*-------------------------------------------------------------------*/
/* Default console tn3270/telnet session TCP keepalive values        */
/*-------------------------------------------------------------------*/
#ifdef _MSVC_
// On Windows Vista and later the default probe count is 10 and
// cannot be changed via setsockopt (only via registry entry).
// On Windows pre-Vista it defaults to 5 probes and can't be
// changed via setsockopt (only via registry entry). Refer to
// function "kasock_init()" in module w32util.c for details.
#define  KEEPALIVE_IDLE_TIME        3   /* Idle time to first probe  */
#define  KEEPALIVE_PROBE_INTERVAL   1   /* Probe timeout value       */
#define  KEEPALIVE_PROBE_COUNT      10  /* Max probe timeouts        */
#else // *nix
// I'm not sure if this is true for all *nix'es, but I know that it
// is for my CentOS 6.4 system: the probe count defaults to '9' and
// apparently cannot be changed via setsockopt. Attempts to change
// it via setsockopt return "succeess" (rc==0), but the value isn't
// actually changed. The value you requested is not honored. A call
// to getsockopt always returns '9'. Other *nix'es may use different
// values or behave differently.
#define  KEEPALIVE_IDLE_TIME        3   /* Idle time to first probe  */
#define  KEEPALIVE_PROBE_INTERVAL   1   /* Probe timeout value       */
#define  KEEPALIVE_PROBE_COUNT      9   /* Max probe timeouts        */
#endif

/*-------------------------------------------------------------------*/
/* Macros for issuing compiler messages                              */
/*-------------------------------------------------------------------*/

#define Q( _s )                 #_s
#define QSTR( _s )              Q( _s )
#define QSTR2( _s1, _s2 )       QSTR( _s1 ## _s2 )
#define QLINE                   __FILE__ "(" QSTR( __LINE__ ) ") : "

#define WARN_LINE               QLINE "warning : "
#define FIXME_LINE              QLINE "FIXME : "
#define TODO_LINE               QLINE "TODO : "

/*-------------------------------------------------------------------*/
/* Macro to suppress compiler "unreferenced variable" warnings       */
/*-------------------------------------------------------------------*/

#define UNREFERENCED(x)         do{}while(0 && x)
#define UNREFERENCED_370(x)     do{}while(0 && x)
#define UNREFERENCED_390(x)     do{}while(0 && x)
#define UNREFERENCED_900(x)     do{}while(0 && x)

/*-------------------------------------------------------------------*/
/* Determine GCC diagnostic pragma support level                     */
/*-------------------------------------------------------------------*/

#if defined( __GNUC__ )
  #define GCC_VERSION ((__GNUC__ * 10000) + (__GNUC_MINOR__ * 100) + __GNUC_PATCHLEVEL__)
  #if GCC_VERSION >= 40200
    #define HAVE_GCC_DIAG_PRAGMA
    #define QPRAGMA( x )                _Pragma( #x )
    #if GCC_VERSION >= 40600
      #define HAVE_GCC_DIAG_PUSHPOP
    #endif
  #endif
#endif

/*-------------------------------------------------------------------*/
/* Same idea, but for issuing a build "#warning ..." message         */
/*-------------------------------------------------------------------*/

#if defined( _MSVC_ )
  #define WARNING( _msg )       __pragma( message( WARN_LINE  _msg ))
#elif defined( __GNUC__ ) && defined( HAVE_GCC_DIAG_PRAGMA )
  #define WARNING( _msg )       QPRAGMA( message( _msg ))
#endif

#ifndef   WARNING
  #define WARNING( _msg )       /* (do nothing) */
#endif


///////////////////////////////////////////////////////////////////////

#define _(x)                    x
#define logmsg                  printf
#define WRMSG(id, s, ...)       logmsg( _(#id s " " id "\n"), ## __VA_ARGS__)
#define get_HSO_errno()         (errno)
#define HSO_errno               get_HSO_errno()
#define close_socket(f)         close(f)

#define HHC02219 "Error in function %s: %s"
#define HHC02321 "This build of Hercules does not support TCP keepalive"
#define HHC02322 "This build of Hercules has only basic TCP keepalive support"
#define HHC02320 "Not all TCP keepalive settings honored"

struct SYSBLK
{
    int  kaidle;
    int  kaintv;
    int  kacnt;
};
typedef struct SYSBLK SYSBLK;

///////////////////////////////////////////////////////////////////////

#endif // _KEEPTEST_H_
=

**Comment by Fish-Git** [comment]
David H. Durgee wrote:
> Fish wrote:
> 
> > Bad test. My fault. My keeptest.c was too simplistic and
> > only enabled "basic" (default) keepalive, which defaults
> > to 2 hours. (Hey I told you I was tired!)
> >
> > Try the attached new/improved version instead. It contains
> > the same logic as in that original Hercules patch I sent
> > you (conkpalv.patch), plus a tiny bit more to report the
> > actual keepalive values being used.
> 
> I built the new program and ran it.  Logs are attached as
> before. Still no sign of keepalive packets.

Don't worry about it. I'VE FOUND THE PROBLEM.

I still have a few more tests to do and then I'll be sending you a new version of "keeptest" as well as a new version of my original "conkpalv.patch" for Hercules.

Hang loose for another day or so. The fix is coming.

-- 
"Fish" (David B. Trout)
Software Development Laboratories
http://www.softdevlabs.com
mail: fish@softdevlabs.com

**Comment by Fish-Git** [comment]
(attachment)
(CC: GitHub)

Dave:

As promised, attached are the fixed versions of my "keeptest" program and an updated Hercules "conkpalv-2.patch".

The conkpalv-2.patch needs to be applied fresh. Reverse the previous patch before applying this new one.

The updated keeptest program now has socket polling logic (pselect call) just like Hercules does. The #define for ERRSET was an experiment to see whether including an error set in the poll (pselect) made any difference or not. (It doesn't.)

The bug was accidentally introduced over a year ago by John P. Hartmann:


    Revision: 9f62223c4879455362b29a6750ff317c0eb79667
    Author: John P. Hartmann <jphartmann@gmail.com>
    Date: 12/24/2013 5:49:46 AM
    Message:
    Make hscutl.c a bit more portable for FreeBSD.
    ----
    Modified: hscutl.c


which I *should* have taken a closer look at much sooner, but hey, hindsight is always 20-20, isn't it? :)

The bug was a simple careless oversight: the call to setsockopt for SO_KEEPALIVE (to enable keepalive on the specified socket) is a socket-level option, not a TCP-level option. Only the calls to adjust the individual keepalive values themselves are TCP-level options.

Thus in the first setsockopt call for SO_KEEPALIVE, it should be specifying "SOL-SOCKET" as the second parameter. When John made his change however, he accidentally changed it to "l_tcp" (SOL_TCP), just like all the other calls. (Oops!)

Now what's interesting about all this (and at the same time, frustrating!) is that the incorrectly coded setsockopt call succeeds anyway! (returns success) I would have expected it to fail. (If it had I might have found the bug a lot sooner!)

Long story short: the problem has been identified and a fix for it now exists. I'm tempted to go ahead and commit my changes anyway (since I tested them six ways to Sunday), but to be safe I'd rather you confirm first that my patch does indeed fix the problem. (I tested on CentOS 6.4 and Windows 7)

The "set_socket_keepalive" code in keeptest.c matches exactly the new (fixed) Hercules code.

Please confirm to me as soon as is convenient to you (preferably via reply to GitHub Issue #44) that my good is good and does indeed resolve your issue, and I will commit the fix to the repository.

Thanks.

-- 
"Fish" (David B. Trout)
Software Development Laboratories
http://www.softdevlabs.com
mail: fish@softdevlabs.com

**Comment by dhdurgee** [comment]
I built and tested the new routine, keepalive function confirmed as operational.  When I get a chance I will apply the patch and test it as well.

**Comment by dhdurgee** [comment]
I had some difficulty sorting out the patching to my hyperion pull, but I beileve I finally got it applied properly.  Unfortunately after running /.autogen.sh, and /.configure when I run the make I am encountering an error as follows:

hsccmd.c: In function 'ctc_cmd':
hsccmd.c:2268:16: error: 'DEBUGPACKET' undeclared (first use in this function)
hsccmd.c:2268:16: note: each undeclared identifier is reported only once for each function it appears in
hsccmd.c: In function 'ptp_cmd':
hsccmd.c:2396:16: error: 'DEBUGPACKET' undeclared (first use in this function)
make[2]: *** [hsccmd.lo] Error 1
make[2]: Leaving directory `/perm/dhdurgee/src/hyperion'
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory `/perm/dhdurgee/src/hyperion'
make: *** [all] Error 2

Perhaps I mangled the retraction of the old patch and applicaton of the new and the error is a result of that.  Let me know if you need more to tell me how to correct this and get hyperion built with the patch applied.

**Comment by Fish-Git** [comment]
[...]
> Unfortunately after running /.autogen.sh, and /.configure
> when I run the make I am encountering an error as follows:

    hsccmd.c: In function 'ctc_cmd':
    hsccmd.c:2268:16: error: 'DEBUGPACKET' undeclared (first use in this function)
    hsccmd.c:2268:16: note: each undeclared identifier is reported only once for each function it appears in
    hsccmd.c: In function 'ptp_cmd':
    hsccmd.c:2396:16: error: 'DEBUGPACKET' undeclared (first use in this function)
    make[2]: *** [hsccmd.lo] Error 1
    make[2]: Leaving directory /perm/dhdurgee/src/hyperion'
    make[1]: *** [all-recursive] Error 1
    make[1]: Leaving directory/perm/dhdurgee/src/hyperion'
    make: *** [all] Error 2

This is neither a problem with anything you did nor a problem with my patch. Rather, it is a problem with Ian Shorter's latest commit. He made some changes related to QETH debugging which included changes to the "ctc" and "ptp" panel commands (which now use the constant mentioned in the above gcc error messages: 'DEBUGPACKET').

Please notify the main hercules-390 list about this so Ian can get it fixed for you.

Thanks.

-- 
"Fish" (David B. Trout)
Software Development Laboratories
http://www.softdevlabs.com
mail: fish@softdevlabs.com

**Comment by Fish-Git** [comment]
Fish wrote:

> This is neither a problem with anything you did
[...]

On second thought... It more than likely *IS* something you did as a result of your attempt to apply my new patch manually.

What you should have done was to REVERSE apply my original patch (i.e. "patch -R -i ../conkpalv.patch") leaving you with your original *unmodified* version of Hyperion that you started with, and then *apply* my new patch to that copy of Hyperion: "patch -i ../conkpalv-2.patch".

But now that you've botched things, the fastest (easiest) and safest way to recover would be to completely discard the Hyperion you've got and simply re-pull (re-download or re-clone or... etc) the *current* version of Hyperion from GitHub and then simply apply my new conkpalv-2.patch to that.

-- 
"Fish" (David B. Trout)
Software Development Laboratories
http://www.softdevlabs.com
mail: fish@softdevlabs.com

**Comment by jphartmann** [comment]
Wouldn't git checkout <module> be sufficient?

**Comment by Fish-Git** [comment]
John P. Hartmann wrote:

> Wouldn't git checkout be sufficient?

Fine. Swell. Whatever. :)

**Comment by dhdurgee** [comment]
Could someone please clarify to a git novice what is necessary to do so?  I did a "make clean" to clear things out and then tried "git checkout" with the following results:

M	cmdtab.h
M	commadpt.c
M	commadpt.h
M	configure.ac
M	console.c
M	hconsts.h
M	hercwind.h
M	hmacros.h
M	hsccmd.c
M	hscutl.c
M	hscutl.h
M	hstdinc.h
M	hstructs.h
M	impl.c
M	msgenu.h
M	version.c
M	w32util.c
M	w32util.h

The above looks like a warning of some sort that the above files were modified.  I assume that I need to specify an additional option to tell git to overwrite these.

**Comment by jphartmann** [comment]
git checkout hsccmd

**Comment by jphartmann** [comment]
Sorry:

git checkout hsccmd.c

**Comment by dhdurgee** [comment]
Does this mean I need to issue a git checkout for each of the 18 files?  Isn't there a way to simply tell it to overwite all of them at once?

**Comment by jphartmann** [comment]
I forget whether it is "git chackout MASTER" or "git checkout HEAD" or whether they do the same.

**Comment by dhdurgee** [comment]
Using "MASTER" results in:

error: pathspec 'MASTER' did not match any file(s) known to git.

Using "HEAD" results in the same behavior as no argument in a list of files prefixed by "M "

**Comment by dhdurgee** [comment]
I am looking at the manpage for git, perhaps the "--bare" option would do what I want?  I am reluctant to try it without further guidance, but the description sounds like what I am after if I properly understand the terminology being used.

**Comment by jphartmann** [comment]
You want "git checkout master" 

That removes any changes you have made.

**Comment by jphartmann** [comment]
You can type "git status" to see what has been changed.

**Comment by dhdurgee** [comment]
Issuing git checkout master says "Already on 'master'" and then the list of files with the "M " prefix.
Issuing git status yields:

# On branch master
# Changes not staged for commit:
#   (use "git add <file>..." to update what will be committed)
#   (use "git checkout -- <file>..." to discard changes in working directory)
#
#	modified:   cmdtab.h
#	modified:   commadpt.c
#	modified:   commadpt.h
#	modified:   configure.ac
#	modified:   console.c
#	modified:   hconsts.h
#	modified:   hercwind.h
#	modified:   hmacros.h
#	modified:   hsccmd.c
#	modified:   hscutl.c
#	modified:   hscutl.h
#	modified:   hstdinc.h
#	modified:   hstructs.h
#	modified:   impl.c
#	modified:   msgenu.h
#	modified:   version.c
#	modified:   w32util.c
#	modified:   w32util.h
#
# Untracked files:
#   (use "git add <file>..." to include in what will be committed)
#
(list of other files in the directory, mostly logs, NICOF and socat related files)

So it appears that "git checkout master" alone will not do it, perhaps if I erase all the files first?

**Comment by jphartmann** [comment]
Did you pull Hercules into just any old directory?  It should have been a dedicated one.

Git checkout overwrites files in your workspace with the version you request.  It does not erase anything.

That said, it is perhaps best that you pull a fresh version and start from scratch.

**Comment by dhdurgee** [comment]
No, I am keeping hyperion in a particular directory.  I also keep hyperion related logs, etc there.

I finally got things to work by issuing git checkout -- <filename> for each of those 18 files and confirmed this was sychronized with the repository via the git status command.  At that point I patched them with the new patch, ran ./autogen.sh followed by ./configure followed by make and make install.

All appears to have worked correctly and I am now testing to confirm the keepalive function is working.

I can now confirm that it IS working, both for 3270 and 2703 connections.

**Comment by Fish-Git** [comment]
dhdurgee wrote:

[...]
> I can now confirm that it IS working, both for 3270
> and 2703 connections.

FINALLY!  :))

Sorry you had so much trouble with git. I use TortoiseGit on Windows which makes using git trivially easy. I'm surprised there isn't a *nix equivalent.

ANYWHO... Now that we've confirmed my fix is good (not that I had any doubt) I'll go ahead and commit it.

Thank you for your incredible patience, David. I really appreciate it.

I will close this issue once I do the commit.

-- 
"Fish" (David B. Trout)
Software Development Laboratories
http://www.softdevlabs.com
mail: fish@softdevlabs.com

**Comment by dhdurgee** [comment]
After my experience with git I think once you have the updates committed I will copy my backups on top of the patched files as opposed to attempting any use of patch to undo it.  I suspect I should then be able to "git pull" to bring your updates in from the repository.  I may do a "git status" first to hopefully confirm it.

**Comment by Fish-Git** [comment]
Closed by commit a2e06edcc9efe25852b3e7737e8c97984855ea5f.
